builder/dispatchers.go

```diff
@@ -288,6 +306,8 @@ func run(b *Builder, args []string, attributes map[string]bool, original string)
 	if err != nil {
 		return err
 	}
+	// revert to original config, we don't persist build time environment
+	b.Config.Env = env
```

Comment:
Yes, this statement rollbacks the b.Config.Env back to original before committing the intermediate container (and it's config). This is needed to prevent build-time environment variables (which are not explicitly defined by ENV) from being persisted in the intermediate images. The defer takes care of the rollback in case of early return.

However, please also see my reply to your other comment below. If we remove this statement we shall be able to keep the build-time environment variables in the intermediate containers, thereby helping with cache lookup in some more cases.